src/coap_mbedtls.c

```diff
@@ -148,6 +161,17 @@ typedef enum coap_enc_method_t {
   COAP_ENC_PKI,
 } coap_enc_method_t;
 
+#ifndef MBEDTLS_2_x
+int coap_rng(void *ctx, unsigned char *buf, size_t len)
```

Comment:
I would prefer not to define this function here and rely on the existing function `coap_prng()`. First of all, this function here is internal, so it should be made static (this would avoid also the [build error](https://github.com/obgm/libcoap/runs/5138665747?check_suite_focus=true#step:7:54). But then, there will be two independent sources of random data in the library: this function and the "official" function `coap_prng()`. If an application decides to set its own different entropy source using `coap_set_prng()` and happens to use Mbed TLS, it ends up with "wrong" random data for the Mbed TLS specific part.

I see why you might want to have this function. In this case, I suggest splitting it off from this PR and provide (configurable) changes against [`coap_prng_default()`](https://github.com/obgm/libcoap/blob/f9c3e13a1802e9d1e9c4ea87763133192990327a/src/coap_prng.c#L52).

Reply:
Hello, @obgm!
Thanks for pointing out these mistakes and suggestions. Does the following patch work?

```patch
diff --git a/src/coap_mbedtls.c b/src/coap_mbedtls.c
index d3dfa72..276023f 100644
--- a/src/coap_mbedtls.c
+++ b/src/coap_mbedtls.c
@@ -148,6 +148,13 @@ typedef enum coap_enc_method_t {
   COAP_ENC_PKI,
 } coap_enc_method_t;

+#ifndef MBEDTLS_2_x
+static int coap_rng(void *ctx, unsigned char *buf, size_t len)
+{
+  return coap_prng(buf, len);
+}
+#endif
+
 static int coap_dgram_read(void *ctx, unsigned char *out, size_t outl)
 {
   ssize_t ret = 0;
diff --git a/src/coap_prng.c b/src/coap_prng.c
index e6ba61d..c3317f2 100644
--- a/src/coap_prng.c
+++ b/src/coap_prng.c
@@ -22,6 +22,10 @@
 #include <stdlib.h>
 #endif /* !HAVE_GETRANDOM */

+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+#include <mbedtls/library/entropy_poll.h>
+#endif /* MBEDTLS_ENTROPY_HARDWARE_ALT */
+
 #if defined(_WIN32)

 errno_t __cdecl rand_s( _Out_ unsigned int* _RandomValue );
@@ -53,6 +57,9 @@ coap_prng_default(void *buf, size_t len) {
 #ifdef HAVE_GETRANDOM
   return getrandom(buf, len, 0);
 #else /* !HAVE_GETRANDOM */
+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+  return mbedtls_hardware_poll(ctx, buf, len, NULL);
+#else
 #if defined(_WIN32)
   return coap_prng_impl(buf,len);
 #else /* !_WIN32 */

```

Reply:
Hello, @obgm!

Thank you for pointing out the mistakes and suggestions. Does the following patch work?
```patch
diff --git a/src/coap_mbedtls.c b/src/coap_mbedtls.c
index d3dfa72..276023f 100644
--- a/src/coap_mbedtls.c
+++ b/src/coap_mbedtls.c
@@ -148,6 +148,13 @@ typedef enum coap_enc_method_t {
   COAP_ENC_PKI,
 } coap_enc_method_t;

+#ifndef MBEDTLS_2_x
+static int coap_rng(void *ctx, unsigned char *buf, size_t len)
+{
+  return coap_prng(buf, len);
+}
+#endif
+
 static int coap_dgram_read(void *ctx, unsigned char *out, size_t outl)
 {
   ssize_t ret = 0;
diff --git a/src/coap_prng.c b/src/coap_prng.c
index e6ba61d..c3317f2 100644
--- a/src/coap_prng.c
+++ b/src/coap_prng.c
@@ -22,6 +22,10 @@
 #include <stdlib.h>
 #endif /* !HAVE_GETRANDOM */

+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+#include <mbedtls/library/entropy_poll.h>
+#endif /* MBEDTLS_ENTROPY_HARDWARE_ALT */
+
 #if defined(_WIN32)

 errno_t __cdecl rand_s( _Out_ unsigned int* _RandomValue );
@@ -53,6 +57,9 @@ coap_prng_default(void *buf, size_t len) {
 #ifdef HAVE_GETRANDOM
   return getrandom(buf, len, 0);
 #else /* !HAVE_GETRANDOM */
+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+  return mbedtls_hardware_poll(ctx, buf, len, NULL);
+#else
 #if defined(_WIN32)
   return coap_prng_impl(buf,len);
 #else /* !_WIN32 */

```

Reply:
Hello, @obgm!

Concerning your comment below, any suggestions on how to split this part off?

PS: I tested this change on the ESP-IDF [master](https://github.com/espressif/esp-idf/commit/5810ed1d043cbc54c9a7616db1e84d99f71f3ccb) with mbedtls-3.1.0 and it worked just fine.